packages/cmake/Modules-Platform-Android.cmake.patch

```diff
@@ -0,0 +1,29 @@
+diff --git a/Modules/Platform/Android.cmake b/Modules/Platform/Android.cmake
+index f08f84176..500fbce03 100644
+--- a/Modules/Platform/Android.cmake
++++ b/Modules/Platform/Android.cmake
+@@ -5,12 +5,20 @@ set(ANDROID 1)
+ # Conventionally Android does not use versioned soname
+ # But in modern versions it is acceptable
+ if(NOT DEFINED CMAKE_PLATFORM_NO_VERSIONED_SONAME)
+-  set(CMAKE_PLATFORM_NO_VERSIONED_SONAME 1)
++  if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Android")
++    set(CMAKE_PLATFORM_NO_VERSIONED_SONAME 0)
++  else()
++    set(CMAKE_PLATFORM_NO_VERSIONED_SONAME 1)
```

Comment:
Tracked down those Swift testsuite errors, turned out to be related to this versioned soname variable, as LLVM uses it for some libraries like libclang. Specifically, the Swift testsuite calls the clang tools c-index-test and c-arcmt-test, which both link against libclang.so, but the dynamic linker complains that it cannot find a versioned symbol from libclang.so when I try to run those executables.

The problem is that this disables versioned sonames, so I've enabled it again by default and checked the generated build.ninja file: they're there again. I've updated this pull just to change this variable and kicked off a build of an updated Swift source snapshot, will report if this fixed the failing swift tests in a couple hours.

@tomty89, I see you pushed this ifdef upstream into CMake, what do you think? I believe support for [this symbol versioning was added with all the linker changes in Android 6](https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md), so this should work for this branch of the package repo.

Reply:
While "this should work", I find the condition you added here confusing (or semantically inappropriate). I would rather simply remove the original condition (downstream in this repo).

Reply:
Or if possible, perhaps we should check the platform/system version here; maybe that could even go upstream. Just that I'm not entirely sure if it's really API 23. (Symbol versioning refers to a different thing AFAIK.)

Reply:
> I find the condition you added here confusing

The idea is to conform with the rest of this pull, ie only change things when the host is Android, which could be upstreamed one day. I could just move the original three lines into the bottom branch with the rpath flag below instead.

> I would rather simply remove the original/parent condition

I kept it on purpose, so this can always be overidden by users who know better or don't want it.

> perhaps we should check the platform/system version here

This pull disables that somewhat complicated checking for the API level, as [that logic assumes the NDK or a standalone toolchain (there's more API checking later in that file)](https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Android-Determine.cmake#L39). Given that someone may not use the default API level and might supply their own, we'd have to tell them to set CMAKE_SYSTEM_VERSION and then check that.

> Just that I'm not entirely sure if it's really API 23.

A ton of this linker support was added with 23, but in any case, that's why I mentioned that we should be good on this branch of this repo, that only supports 24 and up. I'm not sure about how far back this will work either, @xeffyr or anyone know?

> Symbol versioning refers to a different thing AFAIK.

I wasn't entirely clear above: the missing symbols were there in the previous libclang.so, but simply versioned, so I believe they were only unused by the dynamic linker because this versioned soname wasn't there. I haven't dealt with these versioned sonames/symbols before, but from googling a bit, I believe [the soname version is used to determine which versioned symbol to use](https://en.m.wikipedia.org/wiki/Soname), hence the connection.

Reply:
> I kept it on purpose, so this can always be overidden by users who know better or don't want it.

AFAIR removing the condition doesn't prevent that, as the user can still set the variable to 1.

Even if that's not the case, you can still change the original condition/default to meet you need, rather than adding a child condition that doesn't semantically make sense.

Reply:
> AFAIR removing the condition doesn't prevent that, as the user can still set the variable to 1.

No, without that condition you added, it will ignore whatever is passed into CMake by the user and reset it here, just tried it to confirm.

> you can still change the original condition/default to meet you need, rather than adding a child condition that doesn't semantically make sense.

The issue is that there's two ambiguities, one we cannot resolve and would have to add an API for and another we simply don't know right now:
* Whether to use versioned sonames depends on what API the user wants to target, which we cannot know without the user specifying it. 
* We don't know exactly which Android API added support for all the linker features needed for this, and I doubt anyone wants to go back and figure it out.

Given all that, I thought your original condition made sense: let the user decide, but choose a default of turning it off, since that's what most would likely want, particularly when cross-compiling. All this does is use a different default for CMake delivered from this repo to Termux, since API 24 and up almost certainly support it.

I've moved this into the block below now, just because it's shorter though it behaves exactly the same. Maybe that will make more sense to you.

Reply:
Isn't it equivalent to having the condition removed when it is "Android"?

Reply:
Yes, it's undefined by default on an Android host now, which I wasn't sure would work so I kept it outside and explicitly unset it before, but I just ran it and confirmed [this works too](https://gitlab.kitware.com/cmake/cmake/blob/master/Source/cmGeneratorTarget.cxx#L3953).

Reply:
I don't follow. You kept on mentioning the variable being reset/unset, but removing the whole condition block should simply leave it undefined (which should ultimately be equivalent to setting it to 0 by default). Nothing is preventing the user from having the variable set to 1 (or 0) AFAICS.

Reply:
First off, let me emphasize again that this works exactly as before, I just changed it to this because it's now shorter. As for the user overriding it, that still works the same way now, ie we're just setting defaults and the user can change it. The only difference is that now the variable is undefined by default, whereas before I was explicitly setting it to false by default, and as my link to the source shows, both are treated the same.

Reply:
The current approach looks fine to me.

It's just that you said:
> No, without that condition you added, it will ignore whatever is passed into CMake by the user and reset it here, just tried it to confirm.

and then:
> As for the user overriding it, that still works the same way now, 

So I'm lost.

As long as the user change the behavior (from undefined hence false to true) by setting the variable in both cases (Android or not), I'm fine with it.

Reply:
Alright, found the shortest possible version of this patch and pushed it, just stubbing out most of this file now. The rpath variable is already set in Platform/Linux just above and the soversion one stays undefined by default, ran it and checked.

@tomty89, maybe it's unclear because we were not specifying the exact variable or condition we were discussing, but writing a bit more loosely. I was saying that I explicitly set the CMAKE_PLATFORM_NO_VERSIONED_SONAME variable to 0 by default on an Android host before, because I wasn't sure that leaving it undefined would work the same.

Given that, I had to keep your original check to see if that variable was defined, otherwise whatever the user passed in would be overridden. Once I knew that leaving that variable undefined worked the same, I was able to reduce all that logic to just this simple exit stub, which also works the same, including allowing user configuration.